src/realm/sync/client.cpp

```diff
@@ -1564,6 +1497,23 @@ void SessionWrapper::force_close()
     m_sess = nullptr;
     // Everything is being torn down, no need to report connection state anymore
     m_connection_state_change_listener = {};
+
+    // All outstanding wait operations must be canceled
```

Comment:
Moving this from finalize() to force_close() means that in tests we send the notifications when the client is shutdown rather than when the session is abandoned, matching the old behavior of blocking wait for completion or client stop. I think this is clearly correct for tests and should have no effect outside of test code.